products/microsoft_defender_for_endpoints.py

```diff
@@ -0,0 +1,114 @@
+import csv
+import json
+import os
+import time
+import datetime
+from pprint import pprint
+import click
+
+import urllib.request
+import urllib.parse
+import sys
+import requests
+
+
+def build_query(filters):
+    query_base = ''
+
+    for key, value in filters.items():
+        if key == 'days':
+            query_base += f'| where Timestamp > ago({value}d)'
+
+        if key == 'minutes':
+            query_base += f'| where Timestamp > ago({value}m)'
+
+        if key == 'hostname':
+            query_base += f'| where DeviceName contains "{value}"'
+
+        if key == 'username':
+            query_base += f'| where AccountName contains "{value}"'
+    return query_base
+
+def process_search(conn, base, user_query): 
+    
+    results = set()
+
+    url = "https://api.securitycenter.microsoft.com/api/advancedqueries/run"
+    query = "DeviceEvents" + base + user_query + "| project DeviceName, AccountName, ProcessCommandLine, FolderPath"
+
+    query_obj = json.dumps({'Query': query}).encode("utf-8")    
+    headers_obj = {
+        "Authorization":"Bearer " + conn, 
+        "Content-Type":"application/json", 
+        "Accept": 'application/json'
+        }
+   
+    try: 
+        response = requests.post(url, data=query_obj, headers=headers_obj)
+    except KeyboardInterrupt: 
+        click.echo("Caught CTRL-C. Rerun surveyor")
+    except Exception as e: 
+        click.echo(f"There was an exception {e}")
+
+    if response.status_code == 200: 
```

Comment:
🗣️ What happens if the response is not a 200?  Is it caught and displayed some other way?  There's a similar block down below too.

Reply:
If the response isn't a 200 there would be no results to add to the CSV and it should return an empty set. Would it be better to receive an error message or just returning empty for non-successful replies is good enough?

Reply:
Good question, I think that depends on your goals for the update and the general project philosophy.  If there is some non-200 response like a timeout, temporarily busy message, or maybe authorization failures you'll silently get an empty set.  Adding a warning or exiting may help diagnose that at least.

Reply:
@amashinchi-rc - tagging you in because we're talking goals and project philosophy. 

I think it makes sense to add output for error messages if we get them so the end user knows whats going on. Do you have differing thoughts on this?